### PR TITLE
Fix illegal memory access in Triton RMSNorm kernel by casting program_id to int64fix(rms_norm): cast row_idx to int64 in _rms_norm_forward_kernel(# 803)

### PR DESCRIPTION
**Pull request by vvvdwbvvv**
## Testing Done

- Hardware Type: <BLANK>
- [ ] run `make test` to ensure correctness
- [ ] run `make checkstyle` to ensure code style
- [ ] run `make test-convergence` to ensure convergence

### CODE REVIEW

**Comment by vvvdwbvvv** [comment]
@Tcc0403 Test is enclosed in 09b20dc and 9b7c1e0
Add `memory_check_for_test` and `estimate_mem` for calculating safe memory for not OOM before overflow test
Add `test_rms_norm_for_prevent_overflow` for testing the overflow in #803 
<details>
  <summary>Test result</summary>
  
      (venv) (3.11.11) vvvdwbvvv@nccucs108:~/develop/Liger-Kernel$ python3 -m pytest test/transformers/test_rms_norm.py -k test_rms_norm_for_prevent_overflow -k test_rms_norm_for_prevent_overflow -v -rP
    =============================================== test session starts ================================================
    platform linux -- Python 3.11.11, pytest-8.4.1, pluggy-1.6.0 -- /home/vvvdwbvvv/develop/Liger-Kernel/venv/bin/python3
    cachedir: .pytest_cache
    rootdir: /home/vvvdwbvvv/develop/Liger-Kernel
    configfile: pyproject.toml
    collected 44 items / 32 deselected / 12 selected                                                                   
    
    test/transformers/test_rms_norm.py::test_rms_norm_for_prevent_overflow[LlamaRMSNorm-0.0-llama-dtype0-0.0001-1e-06-1-1000000-512] SKIPPED [  8%]
    test/transformers/test_rms_norm.py::test_rms_norm_for_prevent_overflow[LlamaRMSNorm-0.0-llama-dtype0-0.0001-1e-06-1-1000000-4096] SKIPPED [ 16%]
    test/transformers/test_rms_norm.py::test_rms_norm_for_prevent_overflow[LlamaRMSNorm-0.0-llama-dtype1-0.2-0.02-1-1000000-512] PASSED [ 25%]
    test/transformers/test_rms_norm.py::test_rms_norm_for_prevent_overflow[LlamaRMSNorm-0.0-llama-dtype1-0.2-0.02-1-1000000-4096] SKIPPED [ 33%]
    test/transformers/test_rms_norm.py::test_rms_norm_for_prevent_overflow[GemmaRMSNorm-1.0-gemma-dtype0-0.0001-1e-06-1-1000000-512] SKIPPED [ 41%]
    test/transformers/test_rms_norm.py::test_rms_norm_for_prevent_overflow[GemmaRMSNorm-1.0-gemma-dtype0-0.0001-1e-06-1-1000000-4096] SKIPPED [ 50%]
    test/transformers/test_rms_norm.py::test_rms_norm_for_prevent_overflow[GemmaRMSNorm-1.0-gemma-dtype1-0.2-0.02-1-1000000-512] PASSED [ 58%]
    test/transformers/test_rms_norm.py::test_rms_norm_for_prevent_overflow[GemmaRMSNorm-1.0-gemma-dtype1-0.2-0.02-1-1000000-4096] SKIPPED [ 66%]
    test/transformers/test_rms_norm.py::test_rms_norm_for_prevent_overflow[BaseRMSNorm-0.0-none-dtype0-0.0001-1e-06-1-1000000-512] SKIPPED [ 75%]
    test/transformers/test_rms_norm.py::test_rms_norm_for_prevent_overflow[BaseRMSNorm-0.0-none-dtype0-0.0001-1e-06-1-1000000-4096] SKIPPED [ 83%]
    test/transformers/test_rms_norm.py::test_rms_norm_for_prevent_overflow[BaseRMSNorm-0.0-none-dtype1-0.2-0.02-1-1000000-512] PASSED [ 91%]
    test/transformers/test_rms_norm.py::test_rms_norm_for_prevent_overflow[BaseRMSNorm-0.0-none-dtype1-0.2-0.02-1-1000000-4096] SKIPPED [100%]
    
    ================================================= warnings summary =================================================
    venv/lib/python3.11/site-packages/_pytest/config/__init__.py:1474
      /home/vvvdwbvvv/develop/Liger-Kernel/venv/lib/python3.11/site-packages/_pytest/config/__init__.py:1474: PytestConfigWarning: Unknown config option: asyncio_mode
      
        self._warn_or_fail_if_strict(f"Unknown config option: {key}\n")
    
    test/transformers/test_rms_norm.py:80
      /home/vvvdwbvvv/develop/Liger-Kernel/test/transformers/test_rms_norm.py:80: PytestUnknownMarkWarning: Unknown pytest.mark.flaky - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
        @pytest.mark.flaky(reruns=3, reruns_delay=2)
    
    -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
    ====================================================== PASSES ======================================================
    ============================= 3 passed, 9 skipped, 32 deselected, 2 warnings in 2.68s ==============================
</details>

File: test/transformers/test_rms_norm.py [hunk 1]
```diff
@@ -183,3 +183,96 @@ def test_correctness_functional(bs, sl, hd, dtype, atol, rtol, reference, offset
...
+    ],
+)
+def test_rms_norm_for_prevent_overflow(bs, sl, hd, dtype, atol, rtol, reference, offset, casting_mode):
+    if not memory_check_for_test(bs, sl, hd, dtype):
+        required_mem = estimate_mem(bs, sl, hd, dtype)
+        pytest.skip(
+            f"Insufficient GPU memory for test with bs={bs}, sl={sl}, hd={hd}, dtype={dtype}. Required: {required_mem:.2f}GB"
+        )
+    # Create a tensor with large values
+    _tensor = torch.randn(bs, sl, hd, device=device, dtype=dtype) * 1e6
```
Comment (Tcc0403) [inline test/transformers/test_rms_norm.py]: The overflow issue occurs in memory indexing not the tensor values. There's no need to create large values by multiplying 1e6.

File: test/transformers/test_rms_norm.py [hunk 2]
```diff
@@ -183,3 +183,96 @@ def test_correctness_functional(bs, sl, hd, dtype, atol, rtol, reference, offset
...
+        (LlamaRMSNorm, 0.0, "llama"),
+        (GemmaRMSNorm, 1.0, "gemma"),
+        (BaseRMSNorm, 0.0, "none"),
+    ],
+)
+def test_rms_norm_for_prevent_overflow(bs, sl, hd, dtype, atol, rtol, reference, offset, casting_mode):
+    if not memory_check_for_test(bs, sl, hd, dtype):
+        required_mem = estimate_mem(bs, sl, hd, dtype)
+        pytest.skip(
+            f"Insufficient GPU memory for test with bs={bs}, sl={sl}, hd={hd}, dtype={dtype}. Required: {required_mem:.2f}GB"
```
Comment (Tcc0403) [inline test/transformers/test_rms_norm.py:258]: We can use `@pytest.mark.skipif()` decorator instead.

File: test/transformers/test_rms_norm.py [hunk 3]
```diff
@@ -183,3 +183,96 @@ def test_correctness_functional(bs, sl, hd, dtype, atol, rtol, reference, offset
...
+        bytes_per_element = 4
+    elif dtype == torch.bfloat16:
+        bytes_per_element = 2
+
+    # Calculate tensor sizes
+    main_tensor_size = bs * sl * hd * bytes_per_element  # _tensor
+    weight_tensor_size = hd * bytes_per_element  # w
+
+    total_tensors = (
+        2 * main_tensor_size + 2 * main_tensor_size + 2 * main_tensor_size + main_tensor_size + weight_tensor_size
```
Comment (Tcc0403) [inline test/transformers/test_rms_norm.py]: I don't quite understand what each value stands for. Could you explain more?

my guess: (input/ref_input + input_grad/ref_input_grad + ??? + _tensor + RMSNorm weight)

File: test/transformers/test_rms_norm.py [hunk 4]
```diff
@@ -183,3 +183,96 @@ def test_correctness_functional(bs, sl, hd, dtype, atol, rtol, reference, offset
...
+    # Create a tensor with large values
+    _tensor = torch.randn(bs, sl, hd, device=device, dtype=dtype) * 1e6
+
+    h1 = _tensor.clone().requires_grad_(True)
+    h2 = _tensor.clone().requires_grad_(True)
+
+    w = torch.randn(hd, device=device, dtype=dtype)
+
+    y1 = liger_rms_norm(X=h1, W=w, eps=1e-6, offset=offset, casting_mode=casting_mode)
+    y2 = LigerRMSNormFunction.apply(h2, w, 1e-6, offset, casting_mode)
```
Comment (Tcc0403) [inline test/transformers/test_rms_norm.py]: Compare with torch function instead.

File: test/transformers/test_rms_norm.py [hunk 5]
```diff
@@ -183,3 +183,96 @@ def test_correctness_functional(bs, sl, hd, dtype, atol, rtol, reference, offset
...
+)
+@pytest.mark.parametrize(
+    "reference, offset, casting_mode",
+    [
+        (LlamaRMSNorm, 0.0, "llama"),
+        (GemmaRMSNorm, 1.0, "gemma"),
+        (BaseRMSNorm, 0.0, "none"),
+    ],
+)
+def test_rms_norm_for_prevent_overflow(bs, sl, hd, dtype, atol, rtol, reference, offset, casting_mode):
```
Comment (Tcc0403) [inline test/transformers/test_rms_norm.py]:
```suggestion
def test_rms_norm_with_long_seqlen(bs, sl, hd, dtype, atol, rtol, reference, offset, casting_mode):
```

File: test/transformers/test_rms_norm.py [hunk 6]
```diff
@@ -183,3 +183,96 @@ def test_correctness_functional(bs, sl, hd, dtype, atol, rtol, reference, offset
...
+    if not memory_check_for_test(bs, sl, hd, dtype):
+        required_mem = estimate_mem(bs, sl, hd, dtype)
+        pytest.skip(
+            f"Insufficient GPU memory for test with bs={bs}, sl={sl}, hd={hd}, dtype={dtype}. Required: {required_mem:.2f}GB"
+        )
+    # Create a tensor with large values
```
Comment (Tcc0403) [inline test/transformers/test_rms_norm.py]:
```suggestion
    # Create a tensor with long seqlen
```

File: test/transformers/test_rms_norm.py [hunk 7]
```diff
@@ -183,3 +183,96 @@ def test_correctness_functional(bs, sl, hd, dtype, atol, rtol, reference, offset
...
+        2 * main_tensor_size + 2 * main_tensor_size + 2 * main_tensor_size + main_tensor_size + weight_tensor_size
+    )
+
+    # Convert to GB
+    total_memory_gb = total_tensors / (1024**3)
+
+    return total_memory_gb
+
+
+def memory_check_for_test(bs, sl, hd, dtype, margin=3, device_id: int = 0):
```
Comment (Tcc0403) [inline test/transformers/test_rms_norm.py]: I prefer decouple `estimate_mem()` from this function to make it more generalized.

The interface can be `check_enough_memory(estimated_memory, device_id)`, so we can move it to `test_util.py` and reuse it.

**Comment by sfc-gh-sbekman** [comment]
Thank you for working on this @vvvdwbvvv and @Tcc0403, as @winglian - suggested it's probably more than one place to fix.

With this PR the forward doesn't crash, but backward does when using the full model.

```
[rank0]:   File "/code/users/stas/github/Liger-Kernel/src/liger_kernel/ops/rms_norm.py", line 573, in backward
[rank0]:     dX, dW = rms_norm_backward(
[rank0]:   File "/code/users/stas/github/Liger-Kernel/src/liger_kernel/ops/rms_norm.py", line 472, in rms_norm_backward
[rank0]:     _rms_norm_backward_kernel[grid](
[rank0]:   File "/home/yak/.local/lib/python3.10/site-packages/triton/runtime/jit.py", line 347, in <lambda>
[rank0]:     return lambda *args, **kwargs: self.run(grid=grid, warmup=False, *args, **kwargs)
[rank0]:   File "/home/yak/.local/lib/python3.10/site-packages/triton/runtime/jit.py", line 591, in run
[rank0]:     kernel.run(grid_0, grid_1, grid_2, stream, kernel.function, kernel.packed_metadata,
[rank0]:   File "/home/yak/.local/lib/python3.10/site-packages/triton/backends/nvidia/driver.py", line 529, in __call__
[rank0]:     self.launch(gridX, gridY, gridZ, stream, function, self.launch_cooperative_grid, global_scratch, *args)
[rank0]: RuntimeError: Triton Error [CUDA]: an illegal memory access was encountered
```

**Comment by sfc-gh-sbekman** [comment]
To repro, just add backward to my original repro:
```
for idx in range(len(model.model.layers)):
    hidden_states = model.model.layers[idx].input_layernorm(hidden_states)
    z = hidden_states.sum()
    z.backward(retain_graph=True)
```

```
Traceback (most recent call last):
  File "/code/users/stas/github/sf/arctictraining/long-seqlen-fa-liger.py", line 26, in <module>
    z.backward(retain_graph=True)
  File "/home/yak/.local/lib/python3.10/site-packages/torch/_tensor.py", line 648, in backward
    torch.autograd.backward(
  File "/home/yak/.local/lib/python3.10/site-packages/torch/autograd/__init__.py", line 353, in backward
    _engine_run_backward(
  File "/home/yak/.local/lib/python3.10/site-packages/torch/autograd/graph.py", line 824, in _engine_run_backward
    return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: CUDA error: an illegal memory access was encountered
```

this one is even outside of Triton and doesn't match the tb of the full model crash, other than the error itself. So potentially there are at least 2 more problems to fix.

File: src/liger_kernel/ops/rms_norm.py [hunk 8]
```diff
@@ -63,7 +63,7 @@ def _rms_norm_forward_kernel(
     3. https://arxiv.org/pdf/1910.07467
     """
 
-    row_idx = tl.program_id(0)
+    row_idx = tl.program_id(0).to(tl.int64)
```
Comment (sfc-gh-sbekman) [inline src/liger_kernel/ops/rms_norm.py:66]: I was told using int64 indexing will have a performance hit. 

Should this be done conditionally to the size of input? That is only if there is an actual need to do so? Most current use cases will not need this. Though it's likely to change later.

Comment (Tcc0403) [reply src/liger_kernel/ops/rms_norm.py:66]: I agree with conditional casting to avoid hurting performance for normal seqlen use cases. But I think it's worth benchmarking and see how much performance drawback int64 indexing would cause.

**Comment by Tcc0403** [comment]
> With this PR the forward doesn't crash, but backward does when using the full model.

@vvvdwbvvv Let's apply int64 indexing in [backward](https://github.com/linkedin/Liger-Kernel/pull/804/files#diff-73d2c774174032bab7bb8bb965069f1d3345993ad0943b2d50c18d091da70dbeR140) as well

File: src/liger_kernel/ops/rms_norm.py (hunk 8 above)
Comment (sfc-gh-sbekman) [reply src/liger_kernel/ops/rms_norm.py:66]: +1 to benchmark. The performance hint came from pytorch core devs.

Comment (Tcc0403) [reply src/liger_kernel/ops/rms_norm.py:66]: Here's the benchmarking result on H100. 
Input shape: seq_len=2048, hidden_size=2^10 to 2^16 
Time elapsed in ms
```diff
--- bench_int64.txt     2025-07-15 05:53:25.498033987 +0000
+++ bench_no_int64.txt  2025-07-15 05:52:02.547822286 +0000

```
<details>
<summary>forward</summary>

```diff
     "y_values_50": [
-      0.009631999768316746,
-      0.01414399966597557,
-      0.024000000208616257,
-      0.04233599826693535,
-      0.08278399705886841,
-      0.16777600347995758
+      0.010239999741315842,
+      0.014208000153303146,
+      0.023984000086784363,
+      0.042847998440265656,
+      0.08316799998283386,
+      0.16383999586105347
],
"y_values_20": [
-      0.009184000082314014,
-      0.013952000066637993,
-      0.023679999634623528,
-      0.04201599955558777,
-      0.08224000036716461,
-      0.16684800386428833
+      0.009747199714183807,
+      0.01398400031030178,
+      0.02364799939095974,
+      0.04243199899792671,
+      0.08262400329113007,
+      0.1629440039396286
],                                                                                                                                                                                                                
"y_values_80": [                                                                                                                                                                                             
-      0.018649600446224213,
-      0.014399999752640724,
-      0.024320000782608986,
-      0.04265600070357323,
-      0.08339200168848038,
-      0.16883200407028198
+      0.017478399351239224,
+      0.014527999795973301,
+      0.024288000538945198,
+      0.04323200136423111,
+      0.08374399691820145,
+      0.16473600268363953
],
```

</details>

<details>
<summary>backward</summary>

```diff
     "y_values_50": [
-      0.4323199987411499,
-      0.6772480010986328,
-      0.5598400235176086,
-      0.5071679949760437,
-      0.5563840270042419,
-      1.7211520075798035
+      0.5472959876060486,
+      0.5481600165367126,
+      0.5030080080032349,
+      0.5523840188980103,
+      0.563808023929596,
+      0.9408640265464783
     ],
     "y_values_20": [
-      0.3967103898525238,
-      0.5258176088333131,
-      0.5239359736442566,
-      0.4783296048641205,
-      0.5227263808250427,
-      0.983564805984497
+      0.5111104249954224,
+      0.5146560072898865,
+      0.4719231903553009,
+      0.5192127823829651,
+      0.5287231922149658,
+      0.9364543795585631
     ],
     "y_values_80": [
-      0.4874303936958313,
-      1.7327615976333617,
-      0.5751360058784485,
-      0.526809573173523,
-      0.5692800283432007,
-      1.8687167644500735
+      0.5599231839179992,
+      0.5602560043334961,
+      0.5184640288352966,
+      0.5708735942840576,
+      0.5762816071510315,
+      0.9448576092720031
     ],
```

</details>

<details>
<summary>full</summary>

```diff
     "y_values_50": [
-      0.9520800113677979,
-      0.951744019985199,
-      0.9632319808006287,
-      0.9576960206031799,
-      0.9471359848976135,
-      1.145632028579712
+      0.9596160054206848,
+      0.9488319754600525,
+      0.9527040123939514,
+      0.9525439739227295,
+      0.9570240080356598,
+      1.0990079641342163
     ],
     "y_values_20": [
-      0.9387583971023559,
-      0.6872959733009338,
-      0.9491519927978516,
-      0.9404607772827148,
-      0.9338240027427673,
-      1.1422656059265137
+      0.9439552187919618,
+      0.938041603565216,
+      0.9406207799911499,
+      0.9372223854064942,
+      0.942470395565033,
+      1.0947903871536255
     ],
     "y_values_80": [
-      0.9795967936515808,
-      0.9751999974250793,
-      0.991424024105072,
-      0.9854080080986023,
-      0.9739840030670166,
-      1.150015950202942
+      0.9852672100067139,
+      0.9796352148056031,
+      0.9834431767463685,
+      0.9834239959716797,
+      0.9896384000778198,
+      1.1040895938873292
     ],
```

</details>

Comment (Tcc0403) [reply src/liger_kernel/ops/rms_norm.py:66]: The difference isn't obvious, so I think int64 indexing can be a general case.

**Comment by Tcc0403** [comment]
I suggest remove the new added test function and simply add an extremely long seq_len as one of the test cases for orignal tests.

(seq_len, hidden_size) = (2_000_000, 128) is sufficient to repro the issue.

File: test/transformers/test_rms_norm.py [hunk 9]
```diff
@@ -183,3 +183,96 @@ def test_correctness_functional(bs, sl, hd, dtype, atol, rtol, reference, offset
...
+
+def estimate_mem(bs, sl, hd, dtype):
+    """
+    Estimate GPU memory requirement for the test case.
+    """
+    # Get bytes per element for the dtype
+    if dtype == torch.float32:
+        bytes_per_element = 4
+    elif dtype == torch.bfloat16:
+        bytes_per_element = 2
```
Comment (sfc-gh-sbekman) [inline test/transformers/test_rms_norm.py]: automate the process
```suggestion
    bytes_per_element = torch.tensor([], dtype=dtype).element_size()
```

**Comment by sfc-gh-sbekman** [comment]
why do you sometimes use `cast` and other times `to`?
```
src/liger_kernel/ops/grpo_loss.py:    off_b = tl.program_id(0).cast(tl.int64)
src/liger_kernel/ops/cross_entropy.py:    program_id = tl.program_id(0).to(tl.int64)
```

**Comment by sfc-gh-sbekman** [comment]
There are all these remaining w/o casting (many others already have the casting to int64):
```
src/liger_kernel/ops/rms_norm.py:    row_idx = tl.program_id(0) * BLOCK_ROW + tl.arange(0, BLOCK_ROW)
src/liger_kernel/ops/softmax.py:    row_id = tl.program_id(0)
src/liger_kernel/ops/softmax.py:    row_id = tl.program_id(0)
src/liger_kernel/ops/softmax.py:    row_id = tl.program_id(0)
src/liger_kernel/ops/softmax.py:    row_id = tl.program_id(0)
src/liger_kernel/ops/sparsemax.py:    pid_row = tl.program_id(0)
src/liger_kernel/ops/sparsemax.py:    row = tl.program_id(0)
src/liger_kernel/ops/fused_neighborhood_attention.py:    row_id = tl.program_id(0)
src/liger_kernel/ops/fused_neighborhood_attention.py:    batch_head_id = tl.program_id(0)
src/liger_kernel/ops/fused_neighborhood_attention.py:    batch_head_id = tl.program_id(0)
src/liger_kernel/ops/fused_neighborhood_attention.py:    batch_head_id = tl.program_id(0)
src/liger_kernel/ops/fused_neighborhood_attention.py:    batch_head_id = tl.program_id(0)
src/liger_kernel/ops/fused_neighborhood_attention.py:    batch_head_id = tl.program_id(0)
src/liger_kernel/ops/fused_neighborhood_attention.py:    batch_head_id = tl.program_id(0)
src/liger_kernel/ops/rope.py:    pid = tl.program_id(0)
src/liger_kernel/ops/dyt.py:    tl.store(DA + start_row_id * tl.cdiv(N, 512) + tl.program_id(0), da)
src/liger_kernel/ops/experimental/embedding.py:    pid_m = tl.program_id(0)
src/liger_kernel/ops/experimental/embedding.py:    pid_m = tl.program_id(0)
src/liger_kernel/ops/qwen2vl_mrope.py:    pid = tl.program_id(0)
src/liger_kernel/ops/layer_norm.py:    row_idx = tl.program_id(0)
src/liger_kernel/ops/layer_norm.py:    row_block_id = tl.program_id(0)
src/liger_kernel/ops/multi_token_attention.py:    row_block = tl.program_id(0)
src/liger_kernel/ops/multi_token_attention.py:    row_block = tl.program_id(0)
src/liger_kernel/ops/group_norm.py:    batch_idx = tl.program_id(0)
src/liger_kernel/ops/group_norm.py:    batch_idx = tl.program_id(0)
```

**Comment by sfc-gh-sbekman** [comment]
> > With this PR the forward doesn't crash, but backward does when using the full model.
> 
> @vvvdwbvvv Let's apply int64 indexing in [backward](https://github.com/linkedin/Liger-Kernel/pull/804/files#diff-73d2c774174032bab7bb8bb965069f1d3345993ad0943b2d50c18d091da70dbeR140) as well

I added it manually and I no longer have the problem with the rest of the Liger-Kernel - can do a full llama model with long sequence! Yay!

**Comment by sfc-gh-sbekman** [comment]
I spoke too soon: `rope`'s `backward` is now failing.

```
[rank0]:   File "/code/users/stas/github/Liger-Kernel/src/liger_kernel/ops/rope.py", line 238, in backward
[rank0]:     dq, dk = rope_backward(dq, dk, cos, sin)
[rank0]:   File "/code/users/stas/github/Liger-Kernel/src/liger_kernel/ops/rope.py", line 180, in rope_backward
[rank0]:     _triton_rope[(n_row,)](
[rank0]:   File "/home/yak/.local/lib/python3.10/site-packages/triton/runtime/jit.py", line 347, in <lambda>
[rank0]:     return lambda *args, **kwargs: self.run(grid=grid, warmup=False, *args, **kwargs)
[rank0]:   File "/home/yak/.local/lib/python3.10/site-packages/triton/runtime/jit.py", line 591, in run
[rank0]:     kernel.run(grid_0, grid_1, grid_2, stream, kernel.function, kernel.packed_metadata,
[rank0]:   File "/home/yak/.local/lib/python3.10/site-packages/triton/backends/nvidia/driver.py", line 529, in __call__
[rank0]:     self.launch(gridX, gridY, gridZ, stream, function, self.launch_cooperative_grid, global_scratch, *args)
[rank0]: RuntimeError: Triton Error [CUDA]: an illegal memory access was encountered
```

**Comment by Tcc0403** [comment]
Yeah, let's do a thorough review to see if there are any other potential overflow issues in indexing in our codebase.
@vvvdwbvvv

**Comment by sfc-gh-sbekman** [comment]
I applied the same fix to `rope` and overcame the issue. not sure if you just want to do that across the board anywhere that `hidden_states` is involved.

**Comment by sfc-gh-sbekman** [comment]
The 2 fixes I added on top of this PR are:

```
diff --git a/src/liger_kernel/ops/rms_norm.py b/src/liger_kernel/ops/rms_norm.py
index ef851af..af867f4 100644
--- a/src/liger_kernel/ops/rms_norm.py
+++ b/src/liger_kernel/ops/rms_norm.py
@@ -137,7 +137,7 @@ def _rms_norm_backward_kernel(
     dw = sum(dy * (x / RMS)). summation over BxT dimension
     """

-    row_block_id = tl.program_id(0)
+    row_block_id = tl.program_id(0).to(tl.int64)
     row_start = row_block_id * rows_per_program
     row_end = min((row_block_id + 1) * rows_per_program, n_rows)
     col_offsets = tl.arange(0, BLOCK_SIZE)
diff --git a/src/liger_kernel/ops/rope.py b/src/liger_kernel/ops/rope.py
index ebf4702..bd8ded7 100644
--- a/src/liger_kernel/ops/rope.py
+++ b/src/liger_kernel/ops/rope.py
@@ -32,7 +32,7 @@ def _triton_rope(

     # cos size: (1, seq_len, head_dim) or (bsz, seq_len, head_dim)
     # stride: (seq_len * head_dim, head_dim, 1)
-    pid = tl.program_id(0)
+    pid = tl.program_id(0).to(tl.int64)

     # locate start address
     q_ptr = q_ptr + pid * q_row_stride
```

**Comment by vvvdwbvvv** [comment]
@Tcc0403 
Add `mem_get_info` in `test/utils` to count (free, total) memory in GB. 
Add large seqlens case in `test_rmsnorm` and `test_rope` with `pytest.mark.skipif` for testing with VRAM under 40GB. 
Fix `rms_norm.py` and `rope.py` for overflow problem with casting `row_idx` in `rms_norm` and `pid` in `rope` to int64

**Comment by sfc-gh-sbekman** [comment]
I re-tested the PR - good to go for my use case. Thank you.

File: test/utils.py [hunk 10]
```diff
@@ -68,6 +68,42 @@ def get_topk(tensor, k=20):
...
+                "free": free / (1024**3), # transform to GB
+                "total": total / (1024**3), # transform to GB
+            }
+        except Exception as e:
+            info["xpu"] = f"Error: {e}"
+
+    if not info:
+        info["message"] = "No supported device found (neither CUDA nor XPU available)"
+
+    return info
```
Comment (Tcc0403) [inline test/utils.py]: minor: we have a global variable `device`, just use it to conditionally return either `torch.cuda.mem_get_info()` or `torch.xpu.mem_get_info()` to match the torch behavior.

**Comment by Tcc0403** [comment]
After some discussion, we only focus fixing the existing illegal memory access issue in this PR. A dedicated unit test for long seqlen scenarios can be future PRs.

**Review by Tcc0403 — APPROVED** [review]

**Comment by sfc-gh-sbekman** [comment]
Do you have a rough ETA for when you plan to make a new release? (Not rushing you, just asking for your plans)

**Comment by Tcc0403** [comment]
> Do you have a rough ETA for when you plan to make a new release? (Not rushing you, just asking for your plans)

@shimizust

**Comment by sfc-gh-sbekman** [comment]
do you have a timeline please?

I'm just not sure whether I should tell the world to use the bleed Liger-Kernel for this fix or ideally a normal release would be much easier on users.

Thank you!

**Comment by Tcc0403** [comment]
As a outside contributor, I am not able to speak for linkedin and tell you the official release. Personally, I would suggest using the nightly liger. I will try working on it if there are any new bugs being reported, and that's the only thing I can guarantee.

**Comment by sfc-gh-sbekman** [comment]
I totally understand. This is not for me but for the users who may or may not be able to use nightly.

**Comment by shimizust** [comment]
Hi, sorry was occupied last week. Will release a new version by eod. cc @sfc-gh-sbekman @Tcc0403

**Comment by sfc-gh-sbekman** [comment]
Thank you so much, @shimizust

**Comment by shimizust** [comment]
@sfc-gh-sbekman Released version 0.6.1

**Comment by sfc-gh-sbekman** [comment]
thank you very much, Steven!
